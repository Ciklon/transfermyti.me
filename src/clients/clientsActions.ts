import { createAction, createAsyncAction } from "typesafe-actions";
import { Mapping } from "~/allEntities/allEntitiesTypes";
import { ClientsByIdModel } from "./clientsTypes";

export const createClients = createAsyncAction(
  "@clients/CREATE_CLIENTS_IN_WORKSPACE_REQUEST",
  "@clients/CREATE_CLIENTS_IN_WORKSPACE_SUCCESS",
  "@clients/CREATE_CLIENTS_IN_WORKSPACE_FAILURE",
)<void, Record<Mapping, ClientsByIdModel>, void>();

export const fetchClients = createAsyncAction(
  "@clients/FETCH_CLIENTS_IN_WORKSPACE_REQUEST",
  "@clients/FETCH_CLIENTS_IN_WORKSPACE_SUCCESS",
  "@clients/FETCH_CLIENTS_IN_WORKSPACE_FAILURE",
)<void, Record<Mapping, ClientsByIdModel>, void>();

export const pushUpdatesToClients = createAction(
  "@clients/PUSH_UPDATES_TO_CLIENTS",
)<Record<Mapping, ClientsByIdModel>>();

export const flipIsClientIncluded = createAction("@clients/FLIP_IS_INCLUDED")<
  string
>();
