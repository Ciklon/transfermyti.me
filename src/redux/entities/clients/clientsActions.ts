import { createAsyncAction, createStandardAction } from 'typesafe-actions';
import { batchClockifyRequests } from '~/redux/utils';
import {
  apiCreateClockifyClient,
  apiFetchClockifyClients,
  apiFetchTogglClients,
} from '../api/clients';
import {
  showFetchErrorNotification,
  updateInTransferEntity,
} from '~/redux/app/appActions';
import { selectClientsTransferPayloadForWorkspace } from './clientsSelectors';
import { ClockifyClient, TogglClient } from '~/types/clientsTypes';
import { EntityType, ReduxDispatch, ReduxGetState } from '~/types/commonTypes';

export const clockifyClientsFetch = createAsyncAction(
  '@clients/CLOCKIFY_FETCH_REQUEST',
  '@clients/CLOCKIFY_FETCH_SUCCESS',
  '@clients/CLOCKIFY_FETCH_FAILURE',
)<void, ClockifyClient[], void>();

export const togglClientsFetch = createAsyncAction(
  '@clients/TOGGL_FETCH_REQUEST',
  '@clients/TOGGL_FETCH_SUCCESS',
  '@clients/TOGGL_FETCH_FAILURE',
)<void, TogglClient[], void>();

export const clockifyClientsTransfer = createAsyncAction(
  '@clients/CLOCKIFY_TRANSFER_REQUEST',
  '@clients/CLOCKIFY_TRANSFER_SUCCESS',
  '@clients/CLOCKIFY_TRANSFER_FAILURE',
)<void, ClockifyClient[], void>();

export const updateIsClientIncluded = createStandardAction(
  '@clients/UPDATE_IS_INCLUDED',
)<string>();

export const fetchClockifyClients = (workspaceId: string) => async (
  dispatch: ReduxDispatch,
) => {
  dispatch(clockifyClientsFetch.request());
  try {
    const clients = await apiFetchClockifyClients(workspaceId);
    return dispatch(clockifyClientsFetch.success(clients));
  } catch (error) {
    dispatch(showFetchErrorNotification(error));
    return dispatch(clockifyClientsFetch.failure());
  }
};

export const fetchTogglClients = (workspaceId: string) => async (
  dispatch: ReduxDispatch,
) => {
  dispatch(togglClientsFetch.request());
  try {
    const clients = await apiFetchTogglClients(workspaceId);
    return dispatch(togglClientsFetch.success(clients));
  } catch (error) {
    dispatch(showFetchErrorNotification(error));
    return dispatch(togglClientsFetch.failure());
  }
};

export const transferClientsToClockify = (
  togglWorkspaceId: string,
  clockifyWorkspaceId: string,
) => async (dispatch: ReduxDispatch, getState: ReduxGetState) => {
  const state = getState();
  const clientRecordsInWorkspace = selectClientsTransferPayloadForWorkspace(
    state,
    togglWorkspaceId,
  );
  if (clientRecordsInWorkspace.length === 0) return Promise.resolve();

  dispatch(clockifyClientsTransfer.request());

  const onClientRecord = (clientRecord: any) => {
    const transferRecord = { ...clientRecord, type: EntityType.Client };
    dispatch(updateInTransferEntity(transferRecord));
  };

  try {
    const clients = await batchClockifyRequests(
      onClientRecord,
      clientRecordsInWorkspace,
      apiCreateClockifyClient,
      clockifyWorkspaceId,
    );
    return dispatch(clockifyClientsTransfer.success(clients));
  } catch (error) {
    dispatch(showFetchErrorNotification(error));
    return dispatch(clockifyClientsTransfer.failure());
  }
};
