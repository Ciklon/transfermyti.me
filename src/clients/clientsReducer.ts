import { createReducer, ActionType } from "typesafe-actions";
import * as utils from "~/utils";
import { togglTimeEntriesFetch } from "~/timeEntries/timeEntriesActions";
import * as clientsActions from "./clientsActions";
import { EntityGroup, EntityType, ToolName } from "~/common/commonTypes";
import { ReduxStateEntryForTool } from "~/redux/reduxTypes";
import {
  ClockifyClientModel,
  CompoundClientModel,
  TogglClientModel,
} from "./clientsTypes";

type ClientsAction = ActionType<
  typeof clientsActions & typeof togglTimeEntriesFetch
>;

export interface ClientsState {
  readonly clockify: ReduxStateEntryForTool<CompoundClientModel>;
  readonly toggl: ReduxStateEntryForTool<CompoundClientModel>;
  readonly isFetching: boolean;
}

export const initialState: ClientsState = {
  clockify: {
    byId: {},
    idValues: [],
  },
  toggl: {
    byId: {},
    idValues: [],
  },
  isFetching: false,
};

const getSchemaProcessStrategy = (workspaceId: string) => (
  value: ClockifyClientModel | TogglClientModel,
): CompoundClientModel => ({
  id: value.id.toString(),
  name: value.name,
  workspaceId,
  linkedId: null,
  isIncluded: true,
  entryCount: 0,
  memberOf: EntityGroup.Clients,
});

export const clientsReducer = createReducer<ClientsState, ClientsAction>(
  initialState,
)
  .handleAction(
    [
      clientsActions.clockifyClientsFetch.success,
      clientsActions.clockifyClientsTransfer.success,
    ],
    (state, { payload }) => {
      const normalizedState = utils.normalizeState({
        toolName: ToolName.Clockify,
        entityGroup: EntityGroup.Clients,
        entityState: state,
        payload: payload.entityRecords,
        schemaProcessStrategy: getSchemaProcessStrategy(payload.workspaceId),
      });

      const linkedState = utils.linkEntitiesInStateByName(
        EntityGroup.Clients,
        normalizedState,
      );
      return { ...linkedState, isFetching: false };
    },
  )
  .handleAction(
    clientsActions.togglClientsFetch.success,
    (state, { payload }) => {
      const normalizedState = utils.normalizeState({
        toolName: ToolName.Toggl,
        entityGroup: EntityGroup.Clients,
        entityState: state,
        payload: payload.entityRecords,
        schemaProcessStrategy: getSchemaProcessStrategy(payload.workspaceId),
      });
      return { ...normalizedState, isFetching: false };
    },
  )
  .handleAction(
    [
      clientsActions.clockifyClientsFetch.request,
      clientsActions.clockifyClientsTransfer.request,
      clientsActions.togglClientsFetch.request,
    ],
    state => ({
      ...state,
      isFetching: true,
    }),
  )
  .handleAction(
    [
      clientsActions.clockifyClientsFetch.failure,
      clientsActions.clockifyClientsTransfer.failure,
      clientsActions.togglClientsFetch.failure,
    ],
    state => ({
      ...state,
      isFetching: false,
    }),
  )
  .handleAction(clientsActions.flipIsClientIncluded, (state, { payload }) =>
    utils.flipEntityInclusion(state, payload),
  )
  .handleAction(togglTimeEntriesFetch.success, (state, { payload }) =>
    utils.appendEntryCountToState({
      entityType: EntityType.Client,
      toolName: ToolName.Toggl,
      entityState: state,
      timeEntries: payload,
    }),
  );
