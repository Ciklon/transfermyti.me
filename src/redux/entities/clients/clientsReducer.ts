import { getType } from 'typesafe-actions';
import { combineActions, handleActions } from 'redux-actions';
import * as utils from '~/redux/utils';
import { togglTimeEntriesFetch } from '~/redux/entities/timeEntries/timeEntriesActions';
import * as clientsActions from './clientsActions';
import {
  ClockifyClientModel,
  CompoundClientModel,
  EntitiesFetchPayloadModel,
  EntityGroup,
  EntityType,
  ReduxAction,
  ReduxStateEntryForTool,
  TogglClientModel,
  TogglTimeEntryModel,
  ToolName,
} from '~/types';

export interface ClientsState {
  readonly clockify: ReduxStateEntryForTool<CompoundClientModel>;
  readonly toggl: ReduxStateEntryForTool<CompoundClientModel>;
  readonly isFetching: boolean;
}

export const initialState: ClientsState = {
  clockify: {
    byId: {},
    idValues: [],
  },
  toggl: {
    byId: {},
    idValues: [],
  },
  isFetching: false,
};

const getSchemaProcessStrategy = (workspaceId: string) => (
  value: ClockifyClientModel | TogglClientModel,
): CompoundClientModel => ({
  id: value.id.toString(),
  name: value.name,
  workspaceId,
  linkedId: null,
  isIncluded: true,
  entryCount: 0,
  memberOf: EntityGroup.Clients,
});

export const clientsReducer = handleActions(
  {
    [combineActions(
      getType(clientsActions.clockifyClientsFetch.success),
      getType(clientsActions.clockifyClientsTransfer.success),
    )]: (
      state: ClientsState,
      {
        payload: { entityRecords, workspaceId },
      }: ReduxAction<EntitiesFetchPayloadModel<ClockifyClientModel>>,
    ): ClientsState => {
      const normalizedState = utils.normalizeState({
        toolName: ToolName.Clockify,
        entityGroup: EntityGroup.Clients,
        entityState: state,
        payload: entityRecords,
        schemaProcessStrategy: getSchemaProcessStrategy(workspaceId),
      });

      return utils.linkEntitiesInStateByName(
        EntityGroup.Clients,
        normalizedState,
      );
    },

    [getType(clientsActions.togglClientsFetch.success)]: (
      state: ClientsState,
      {
        payload: { entityRecords, workspaceId },
      }: ReduxAction<EntitiesFetchPayloadModel<TogglClientModel>>,
    ): ClientsState =>
      utils.normalizeState({
        toolName: ToolName.Toggl,
        entityGroup: EntityGroup.Clients,
        entityState: state,
        payload: entityRecords,
        schemaProcessStrategy: getSchemaProcessStrategy(workspaceId),
      }),

    [combineActions(
      getType(clientsActions.clockifyClientsFetch.request),
      getType(clientsActions.clockifyClientsTransfer.request),
      getType(clientsActions.togglClientsFetch.request),
    )]: (state: ClientsState): ClientsState => ({
      ...state,
      isFetching: true,
    }),

    [combineActions(
      getType(clientsActions.clockifyClientsFetch.success),
      getType(clientsActions.clockifyClientsFetch.failure),
      getType(clientsActions.clockifyClientsTransfer.success),
      getType(clientsActions.clockifyClientsTransfer.failure),
      getType(clientsActions.togglClientsFetch.success),
      getType(clientsActions.togglClientsFetch.failure),
    )]: (state: ClientsState): ClientsState => ({
      ...state,
      isFetching: false,
    }),

    [getType(clientsActions.flipIsClientIncluded)]: (
      state: ClientsState,
      { payload: clientId }: ReduxAction<string>,
    ): ClientsState => utils.flipEntityInclusion(state, clientId),

    [getType(togglTimeEntriesFetch.success)]: (
      state: ClientsState,
      { payload: timeEntries }: ReduxAction<Array<TogglTimeEntryModel>>,
    ) =>
      utils.appendEntryCountToState({
        entityType: EntityType.Client,
        toolName: ToolName.Toggl,
        entityState: state,
        timeEntries,
      }),
  },
  initialState,
);
