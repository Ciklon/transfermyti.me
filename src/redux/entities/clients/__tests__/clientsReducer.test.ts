import cases from 'jest-in-case';
import * as clientsActions from '../clientsActions';
import clientsReducer, { initialState } from '../clientsReducer';

const clockifyClientsPayload = [
  {
    id: '2FZn1Y9R5M4nnMGV9K32xJ',
    name: 'Client A',
    workspaceId: 'aZJSPbcPw1cpdfYmtHM7za',
  },
  {
    id: 'ir4aN5VF7pdfDrgWjGCxeR',
    name: 'Client B',
    workspaceId: 'aZJSPbcPw1cpdfYmtHM7za',
  },
  {
    id: 'uABwBW6XWsWLRzRtKY1gow',
    name: 'Client C',
    workspaceId: 'aZJSPbcPw1cpdfYmtHM7za',
  },
];

const togglClientsPayload = [
  {
    id: 37389358,
    wid: 1738517,
    name: 'Client A',
    at: '2018-02-18T16:54:02+00:00',
  },
  {
    id: 19521509,
    wid: 1738517,
    name: 'Client B',
    at: '2016-12-06T03:48:38+00:00',
  },
  {
    id: 21773956,
    wid: 1738517,
    name: 'Client C',
    at: '2017-04-26T14:18:07+00:00',
  },
];

describe('clientsReducer', () => {
  cases(
    'state matches snapshot for fetch actions',
    options => {
      const result = clientsReducer(
        initialState,
        options.action(options.payload),
      );

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: 'when clockifyClientsFetch.request is dispatched',
        action: clientsActions.clockifyClientsFetch.request,
        payload: undefined,
      },
      {
        name: 'when clockifyClientsFetch.success is dispatched',
        action: clientsActions.clockifyClientsFetch.success,
        payload: clockifyClientsPayload,
      },
      {
        name: 'when clockifyClientsFetch.failure is dispatched',
        action: clientsActions.clockifyClientsFetch.failure,
        payload: undefined,
      },
      {
        name: 'when clockifyClientsTransfer.request is dispatched',
        action: clientsActions.clockifyClientsTransfer.request,
        payload: undefined,
      },
      {
        name: 'when clockifyClientsTransfer.success is dispatched',
        action: clientsActions.clockifyClientsTransfer.success,
        payload: clockifyClientsPayload,
      },
      {
        name: 'when clockifyClientsTransfer.failure is dispatched',
        action: clientsActions.clockifyClientsTransfer.failure,
        payload: undefined,
      },
      {
        name: 'when togglClientsFetch.request is dispatched',
        action: clientsActions.togglClientsFetch.request,
        payload: undefined,
      },
      {
        name: 'when togglClientsFetch.success is dispatched',
        action: clientsActions.togglClientsFetch.success,
        payload: togglClientsPayload,
      },
      {
        name: 'when togglClientsFetch.failure is dispatched',
        action: clientsActions.togglClientsFetch.failure,
        payload: undefined,
      },
    ],
  );

  cases(
    'sets isFetching = true',
    options => {
      const result = clientsReducer(
        { ...initialState, isFetching: false },
        options.action(),
      );

      expect(result.isFetching).toBe(true);
    },
    [
      {
        name: 'when clockifyClientsFetch.request is dispatched',
        action: clientsActions.clockifyClientsFetch.request,
      },
      {
        name: 'when clockifyClientsTransfer.request is dispatched',
        action: clientsActions.clockifyClientsTransfer.request,
      },
      {
        name: 'when togglClientsFetch.request is dispatched',
        action: clientsActions.togglClientsFetch.request,
      },
    ],
  );

  cases(
    'sets isFetching = false',
    options => {
      const result = clientsReducer(
        { ...initialState, isFetching: true },
        options.action(options.payload),
      );

      expect(result.isFetching).toBe(false);
    },
    [
      {
        name: 'when clockifyClientsFetch.success is dispatched',
        action: clientsActions.clockifyClientsFetch.success,
        payload: clockifyClientsPayload,
      },
      {
        name: 'when clockifyClientsFetch.failure is dispatched',
        action: clientsActions.clockifyClientsFetch.failure,
        payload: undefined,
      },
      {
        name: 'when clockifyClientsTransfer.success is dispatched',
        action: clientsActions.clockifyClientsTransfer.success,
        payload: clockifyClientsPayload,
      },
      {
        name: 'when clockifyClientsTransfer.failure is dispatched',
        action: clientsActions.clockifyClientsTransfer.failure,
        payload: undefined,
      },
      {
        name: 'when togglClientsFetch.success is dispatched',
        action: clientsActions.togglClientsFetch.success,
        payload: togglClientsPayload,
      },
      {
        name: 'when togglClientsFetch.failure is dispatched',
        action: clientsActions.togglClientsFetch.failure,
        payload: undefined,
      },
    ],
  );

  test('flips isIncluded when flipIsClientIncluded is dispatched', () => {
    const CLIENT_ID = 'ir4aN5VF7pdfDrgWjGCxeR';
    const testState = {
      ...initialState,
      toggl: {
        ...initialState.toggl,
        byId: {
          [CLIENT_ID]: {
            entryCount: 0,
            id: CLIENT_ID,
            isIncluded: false,
            linkedId: '',
            name: 'Client B',
            workspaceId: 'aZJSPbcPw1cpdfYmtHM7za',
          },
        },
      },
    };
    const result = clientsReducer(
      testState,
      clientsActions.flipIsClientIncluded(CLIENT_ID),
    );

    expect(result.toggl.byId[CLIENT_ID].isIncluded).toBe(true);
  });
});
