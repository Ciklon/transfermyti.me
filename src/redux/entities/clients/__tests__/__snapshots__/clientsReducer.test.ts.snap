// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsFetch.failure is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsFetch.request is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": true,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsFetch.success is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {
      "2FZn1Y9R5M4nnMGV9K32xJ": Object {
        "entryCount": 0,
        "id": "2FZn1Y9R5M4nnMGV9K32xJ",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client A",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
      "ir4aN5VF7pdfDrgWjGCxeR": Object {
        "entryCount": 0,
        "id": "ir4aN5VF7pdfDrgWjGCxeR",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client B",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
      "uABwBW6XWsWLRzRtKY1gow": Object {
        "entryCount": 0,
        "id": "uABwBW6XWsWLRzRtKY1gow",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client C",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
    },
    "idValues": Array [
      "2FZn1Y9R5M4nnMGV9K32xJ",
      "ir4aN5VF7pdfDrgWjGCxeR",
      "uABwBW6XWsWLRzRtKY1gow",
    ],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsTransfer.failure is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsTransfer.request is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": true,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when clockifyClientsTransfer.success is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {
      "2FZn1Y9R5M4nnMGV9K32xJ": Object {
        "entryCount": 0,
        "id": "2FZn1Y9R5M4nnMGV9K32xJ",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client A",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
      "ir4aN5VF7pdfDrgWjGCxeR": Object {
        "entryCount": 0,
        "id": "ir4aN5VF7pdfDrgWjGCxeR",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client B",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
      "uABwBW6XWsWLRzRtKY1gow": Object {
        "entryCount": 0,
        "id": "uABwBW6XWsWLRzRtKY1gow",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client C",
        "workspaceId": "aZJSPbcPw1cpdfYmtHM7za",
      },
    },
    "idValues": Array [
      "2FZn1Y9R5M4nnMGV9K32xJ",
      "ir4aN5VF7pdfDrgWjGCxeR",
      "uABwBW6XWsWLRzRtKY1gow",
    ],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when togglClientsFetch.failure is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when togglClientsFetch.request is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": true,
  "toggl": Object {
    "byId": Object {},
    "idValues": Array [],
  },
}
`;

exports[`clientsReducer state matches snapshot for fetch actions when togglClientsFetch.success is dispatched 1`] = `
Object {
  "clockify": Object {
    "byId": Object {},
    "idValues": Array [],
  },
  "isFetching": false,
  "toggl": Object {
    "byId": Object {
      "19521509": Object {
        "entryCount": 0,
        "id": "19521509",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client B",
        "workspaceId": "1738517",
      },
      "21773956": Object {
        "entryCount": 0,
        "id": "21773956",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client C",
        "workspaceId": "1738517",
      },
      "37389358": Object {
        "entryCount": 0,
        "id": "37389358",
        "isIncluded": true,
        "linkedId": null,
        "memberOf": "clients",
        "name": "Client A",
        "workspaceId": "1738517",
      },
    },
    "idValues": Array [
      "37389358",
      "19521509",
      "21773956",
    ],
  },
}
`;
